### PR TITLE
Implement `ecdsa_to_eth_address()` and remove `eth_compatibility` crate

### LINKED ISSUE #1068: Remove `libsecp256k1` dependency of `ink_eth_compatibility`
We need the `libsecp256k1` dependency only in the `to_eth_address` function, for getting the uncompressed 65-byte public key from the ECDSA public key.

Ways to remove the dep would be to either add a new API function in the `contracts` pallet, find another no-std compatible crate, or extract the functionality from either `libsecp256k1` or `secp256k1` into `ink_eth_compatibility`.

### PR DESCRIPTION
**Pull request by agryaznov**
- Follow up to [substrate#11087](https://github.com/paritytech/substrate/pull/11087)
- Resolves #1068 

Ported the following function from `eth_compatibility` crate to `ink_env`:
```rust
pub fn ecdsa_to_eth_address(pubkey: &[u8; 33], output: &mut [u8; 20]) -> Result<()>
``` 
`eth_compatibility` crate removed as having no reason to exist anymore

### CODE REVIEW

**Comment by athei** [comment]
The function shouldn't be removed. It should be replaced with a call to the new seal function.

**Comment by agryaznov** [comment]
I was just thinking on removal the whole `ink_eth_compatibility` crate and adding thus seal function just like any other seal function into `ink_env` and `ink_lang`

File: crates/lang/src/env_access.rs [hunk 1]
```diff
@@ -831,10 +830,17 @@ where
...
-    ) -> Result<ECDSAPublicKey> {
+    ) -> Result<[u8; 33]> {
         let mut output = [0; 33];
         ink_env::ecdsa_recover(signature, message_hash, &mut output)
-            .map(|_| output.into())
+            .map(|_| output)
+            .map_err(|_| Error::EcdsaRecoveryFailed)
+    }
+
+    pub fn ecdsa_to_eth_address(self, pubkey: &[u8; 33]) -> Result<[u8; 20]> {
```
Comment (athei) [inline crates/lang/src/env_access.rs:840]: I'd be curious of LLVM is able to do copy elision of the return value here. It is not guaranteed by Rust (yet).

**Review by athei — APPROVED** [review]

File: crates/env/src/api.rs [hunk 2]
```diff
@@ -500,6 +500,13 @@ pub fn ecdsa_recover(
     })
 }
 
+/// Returns Ethereum address from the ECDSA compressed public key.
```
Comment (cmichi) [inline crates/env/src/api.rs:503]:
```suggestion
/// Returns an Ethereum address from the ECDSA compressed public key.
///
/// # Example
///
/// ```
/// let pub_key = [
///     2, 121, 190, 102, 126, 249, 220, 187, 172, 85, 160,  98, 149, 206, 135, 11,
///     7,   2, 155, 252, 219,  45, 206,  40, 217, 89, 242, 129,  91,  22, 248, 23,
///     152,
/// ];
///
/// let EXPECTED_ETH_ADDRESS = [
///     126, 95, 69, 82, 9, 26, 105, 18, 93, 93, 252, 183, 184, 194, 101, 144, 41, 57, 91, 223
/// ];
///
/// let mut output = [0; 33];
/// ink_env::ecsda_to_eth_address(&pubkey, &mut output)
/// assert_eq!(output, Ok(EXPECTED_ETH_ADDRESS));
/// ```
///
/// # Errors
```

The code will likely not work out of the box, I just copied it from the removed crate below and modified it a bit.
Please add the error cases there as well.

File: crates/env/src/backend.rs [hunk 3]
```diff
@@ -249,6 +249,14 @@ pub trait EnvBackend {
         output: &mut [u8; 33],
     ) -> Result<()>;
 
+    /// Retrieves Ethereum address from the ECDSA compressed `pubkey`,
```
Comment (cmichi) [inline crates/env/src/backend.rs:252]:
```suggestion
    /// Retrieves an Ethereum address from the ECDSA compressed `pubkey`,
```

File: crates/eth_compatibility/src/lib.rs [hunk 4]
```diff
@@ -1,150 +0,0 @@
...
-    ///     152,
-    /// ].into();
-    ///
-    /// const EXPECTED_ACCOUNT_ID: [u8; 32] = [
-    ///     41, 117, 241, 210, 139, 146, 182, 232,  68, 153, 184, 59,   7, 151, 239, 82,
-    ///     53,  85,  62, 235, 126, 218, 160, 206, 162,  67, 193, 18, 140,  47, 231, 55,
-    /// ];
-    ///
-    /// assert_eq!(pub_key.to_default_account_id(), EXPECTED_ACCOUNT_ID.into());
-    pub fn to_default_account_id(
```
Comment (cmichi) [inline crates/eth_compatibility/src/lib.rs:140]: @athei Hmm, do you see any way how we could keep this function?

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 ⚠️ **The ink! `master` is ahead of your branch, this might skew the comparison data below.**

These are the results when building the `examples/*` contracts from this branch with `cargo-contract 1.2.0-9f9b9b3` and comparing them to ink! `master`: 

|                          | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:--------------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`            |                  |            | 1.03 K               |                |
| `adder`                  |                  |            | 2.13 K               |                |
| `contract-introspection` |                  |            | 2.37 K               |                |
| `contract-terminate`     |                  |            | 0.94 K               | 213_894        |
| `contract-transfer`      |                  |            | 8.13 K               | 13_894         |
| `delegate-calls`         |                  |            | 2.95 K               | 14_415         |
| `delegator`              |                  |            | 6.37 K               | 44_665         |
| `dns`                    |                  |            | 8.84 K               | 41_682         |
| `erc1155`                |                  |            | 17.34 K              | 83_364         |
| `erc20`                  |                  |            | 8.49 K               | 41_682         |
| `erc721`                 |                  |            | 11.81 K              | 111_152        |
| `flipper`                |                  |            | 1.31 K               | 13_894         |
| `forward-calls`          |                  |            | 2.88 K               | 28_474         |
| `incrementer`            |                  |            | 1.21 K               |                |
| `multisig`               |                  |            | 25.33 K              | 90_231         |
| `rand-extension`         |                  |            | 3.92 K               | 13_894         |
| `seal-code-hash`         |                  |            | 1.59 K               |                |
| `set-code-hash`          |                  |            | 1.58 K               | 27_788         |
| `subber`                 |                  |            | 2.15 K               |                |
| `trait-erc20`            |                  |            | 8.76 K               | 41_682         |
| `trait-flipper`          |                  |            | 1.00 K               | 13_894         |
| `trait-incrementer`      |                  |            | 1.19 K               | 27_788         |
| `updated-incrementer`    |                  |            | 9.69 K               |                |
| `upgradeable-flipper`    |                  |            | 1.55 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/191468) | Last update: Mon May  2 17:46:25 CEST 2022

File: crates/eth_compatibility/src/lib.rs (hunk 4 above)
Comment (agryaznov) [reply crates/eth_compatibility/src/lib.rs:140]: I think we could make this as a [trait extension](https://github.com/paritytech/substrate/pull/11087/files#diff-9237527ee4beab4bda27f69f53a62a4dc122bb00d92ab22d21701e6d21813b0dR35) for `Public` in `frame_support`, and then add a `seal_` function to `pallet_contracts` API, just like it's done for `to_eth_address()`. Question is: is this function really needed though, what are use cases for it?

Comment (cmichi) [reply crates/eth_compatibility/src/lib.rs:140]: >Question is: is this function really needed though, what are use cases for it?

@xgreenx Do you have something to add here? Is this function maybe used by OpenBrush?

Comment (athei) [reply crates/eth_compatibility/src/lib.rs:140]: This function doesn't depend on any async cryptography. It just hashes some bytes which relies on pallet contracts. We could just copy it to `ink_env`. However, I would just remove it as hash functions are already exposed.

File: crates/env/src/api.rs [hunk 5]
```diff
@@ -599,3 +628,15 @@ where
 pub fn set_code_hash(code_hash: &[u8; 32]) -> Result<()> {
     <EnvInstance as OnInstance>::on_instance(|instance| instance.set_code_hash(code_hash))
 }
+
+/// Returns the default Substrate's `AccountId` (`\[u8;32\]`) from the ECDSA compressed public key.
```
Comment (cmichi) [inline crates/env/src/api.rs:632]:
```suggestion
/// Returns the default Substrate `AccountId` (`[u8;32]`) from the ECDSA compressed public key.
```

File: crates/env/src/api.rs [hunk 6]
```diff
@@ -599,3 +628,15 @@ where
...
+/// It hashes the compressed public key with the `blake2b_256` algorithm like in substrate.
+///
+/// # Note
+///
+/// This function implies a standart `AccountId` type which is `\[u8;32\]`.
```
Comment (cmichi) [inline crates/env/src/api.rs:637]:
```suggestion
/// This function implies a standard `AccountId` type which is `[u8; 32]`.
```

File: crates/env/src/backend.rs [hunk 7]
```diff
@@ -253,6 +254,14 @@ pub trait EnvBackend {
         output: &mut [u8; 33],
     ) -> Result<()>;
 
+    /// Retrieves an Ethereum address from the ECDSA compressed `pubkey`,
```
Comment (cmichi) [inline crates/env/src/backend.rs:257]:
```suggestion
    /// Retrieves an Ethereum address from the ECDSA compressed `pubkey`
```

File: crates/env/src/backend.rs [hunk 8]
```diff
@@ -295,6 +304,14 @@ pub trait EnvBackend {
     ///
     /// - If the supplied `code_hash` cannot be found on-chain.
     fn set_code_hash(&mut self, code_hash: &[u8]) -> Result<()>;
+
+    /// Returns the default Substrate's `AccountId` ([u8;32]) from the ECDSA compressed public key.
```
Comment (cmichi) [inline crates/env/src/backend.rs:308]:
```suggestion
    /// Returns the default Substrate `AccountId` (`[u8; 32]`) from the ECDSA compressed public key.
```

File: crates/lang/src/env_access.rs [hunk 9]
```diff
@@ -831,10 +831,59 @@ where
...
+    ///         145,
+    ///     ];
+    ///      let EXPECTED_ETH_ADDRESS = [
+    ///         9, 35, 29, 167, 177, 154, 1, 111, 158, 87, 109, 35, 177, 98, 119, 6, 47,
+    ///         77, 70, 168,
+    ///     ];
+    ///     let output = self
+    ///         .env()
+    ///         .ecdsa_to_eth_address(&pub_key)
+    ///         .expect("Should get ETH address from compressed public key");
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:872]:
```suggestion
    ///         .expect("must return an Ethereum address for the compressed public key");
```

File: crates/lang/src/env_access.rs [hunk 10]
```diff
@@ -965,4 +1014,14 @@ where
     pub fn own_code_hash(self) -> Result<E::Hash> {
         ink_env::own_code_hash::<E>()
     }
+
+    /// Returns the default Substrate's `AccountId` from the ECDSA compressed public key.
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:1018]:
```suggestion
    /// Returns the default Substrate `AccountId` from the ECDSA compressed public key.
```

File: crates/env/src/api.rs [hunk 11]
```diff
@@ -511,6 +512,34 @@ pub fn ecdsa_recover(
...
+///     77, 70, 168,
+/// ];
+/// let mut output = [0; 20];
+/// ink_env::ecdsa_to_eth_address(&pub_key, &mut output);
+/// assert_eq!(output, EXPECTED_ETH_ADDRESS);
+/// ```
+///
+/// # Errors
+///
+/// - if cannot retrieve ECDSA public key from the provided input
```
Comment (cmichi) [inline crates/env/src/api.rs:536]:
```suggestion
/// - If the ECDSA public key cannot be recovered from the provided public key.
```

**Comment by agryaznov** [comment]
After further discussion it was finally decided to remove `ecdsa_to_default_account_id()`

**Review by cmichi — APPROVED** [review]

**Pull request merged**